tests/phpunit/tests/image/functions.php

```diff
@@ -801,6 +801,92 @@ public function test_pdf_preview_doesnt_overwrite_existing_jpeg() {
 		}
 	}
 
+	/**
+	 * Test short-circuiting of additional image source generation for mime type.
+	 */
+	public function test_wp_content_pre_generate_additional_image_source() {
+		$temp_dir = get_temp_dir();
+		$img_path = $temp_dir . 'test.jpg';
+		copy( DIR_TESTDATA . '/images/33772.jpg', $img_path );
+
+		$editor = wp_get_image_editor( $img_path );
+
+		if ( is_wp_error( $editor ) ) {
+			$this->markTestSkipped( $editor->get_error_message() );
+		}
+
+		add_filter( 'wp_content_pre_generate_additional_image_source', array( $this, 'filter_pre_generate_additional_image_source' ), 10, 5 );
+
+		$attachment_id = $this->factory->attachment->create_object(
+			$img_path,
+			0,
+			array(
+				'post_mime_type' => 'image/jpg',
+			)
+		);
+
+		$metadata = wp_generate_attachment_metadata( $attachment_id, $img_path );
+
+		foreach ( $metadata['sizes'] as $size_name => $size_meta ) {
+			$this->assertSame( "filter-image-{$size_meta['width']}x{$size_meta['height']}.webp", $size_meta['sources']['image/webp']['file'] );
+		}
+
+		remove_filter( 'wp_content_pre_generate_additional_image_source', array( $this, 'filter_pre_generate_additional_image_source' ), 10, 5 );
+	}
+
+	public function filter_pre_generate_additional_image_source( $size_meta, $file, $attachment_id, $size_name, $mime_type ) {
+		if ( 'image/webp' === $mime_type ) {
+			$size_meta['sources'][ $mime_type ] = array(
+				'file'     => "filter-image-{$size_meta['width']}x{$size_meta['height']}.webp",
+				'filesize' => 1000,
+			);
+		}
+
+		return $size_meta;
+	}
+
+	/**
+	 * Test replacing of image filename in the desired context.
+	 */
+	public function test_wp_content_pre_replace_additional_image_source_for_size_thumbnail() {
+		$temp_dir = get_temp_dir();
+		$img_path = $temp_dir . 'test.jpg';
+		copy( DIR_TESTDATA . '/images/33772.jpg', $img_path );
+
+		$editor = wp_get_image_editor( $img_path );
+
+		if ( is_wp_error( $editor ) ) {
+			$this->markTestSkipped( $editor->get_error_message() );
+		}
+
+		add_filter( 'wp_content_pre_replace_additional_image_source', array( $this, 'filter_pre_replace_additional_image_source' ), 10, 6 );
```

Comment:
Add another assertion here before adding the filter where the returned image still has the `.jpg` extension. This demonstrates that the filter is what is adding the image.